File: programmers/2018 KAKAO BLIND RECRUITMENT/[1차] 다트 게임(Level1).py
'''
    Code written on Aug 26, 2020
    출처: 프로그래머스 코딩 테스트 연습, https://programmers.co.kr/learn/challenges
    2018 KAKAO BLIND RECRUITMENT: [1차] 다트 게임(Level1)
    link: https://programmers.co.kr/learn/courses/30/lessons/17682
'''
def solution(dartResult):
    domain = {'S' : 1, 'D' : 2, 'T' : 3}
    before = now = total = 0
    num = ''
    for dart in dartResult:
        if dart.isdigit():
            num += dart
        elif dart in domain:
            total += before
            before = now
            now = int(num) ** domain[dart]
            num = ''
        elif dart == '#':
            now *= -1
        else:
            before *= 2
            now *= 2
    total += before + now
    return total

File: programmers/2018 KAKAO BLIND RECRUITMENT/[1차] 다트 게임(Level1).java
/*
    Code written on Dec 2, 2020
    출처: 프로그래머스 코딩 테스트 연습, https://programmers.co.kr/learn/challenges
    2018 KAKAO BLIND RECRUITMENT: [1차] 다트 게임(Level1)
    link: https://programmers.co.kr/learn/courses/30/lessons/17682
*/
import java.util.Arrays;
import java.util.HashMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class Solution {
    public int solution(String dartResult) {
        Matcher matcher = Pattern.compile("[0-9]+|[SDT]|[*#]").matcher(dartResult);
        HashMap<String, Integer> exp = new HashMap<>() {{
            put("S", 1);
            put("D", 2);
            put("T", 3);
        }};
        String str;
        int[] scores = new int[4];
        int round = 0;
        
        while (matcher.find()) {
            str = matcher.group();
            if (exp.containsKey(str)) {
                scores[round] = (int) Math.pow(scores[round], exp.get(str));
            } else if (str.equals("*")){
                scores[round-1] *= 2;
                scores[round] *= 2;
            } else if (str.equals("#")) {
                scores[round] *= -1;
            } else {
                scores[++round] = Integer.parseInt(str);
            }
        }
        return Arrays.stream(scores).sum();
    }
}
